components/fs/fat/event.c

```diff
@@ -0,0 +1,281 @@
+        LOG_FATFS("%02x ", buffer[i]); // Print each byte in hexadecimal
+    }
+    LOG_FATFS("\n");
+}
+
+_Static_assert(BLK_QUEUE_SIZE_CLI_FATFS >= WORKER_THREAD_NUM, 
+    "The size of queue between fs and blk should be at least the size of WORKER_THREAD_NUM");
+
+void init(void) {
+    // Init the block device queue
+    // Have to make sure who initialize this SDDF queue
+    blk_queue_init(blk_queue_handle, blk_request, blk_response, BLK_QUEUE_SIZE_CLI_FATFS);
+    /*
+       This part of the code is for setting up the thread pool by
+       assign stacks and size of the stack to the pool
+    */
+    uint64_t stack[WORKER_THREAD_NUM];
+    stack[0] = worker_thread_stack_one;
+    stack[1] = worker_thread_stack_two;
+    stack[2] = worker_thread_stack_three;
+    stack[3] = worker_thread_stack_four;
+
+    // Init thread pool
+    microkit_cothread_init(&co_controller_mem,
+                            WORKER_THREAD_STACKSIZE,
+                            stack[0],
+                            stack[1],
+                            stack[2],
+                            stack[3]);
+    for (uint32_t i = 0; i < (WORKER_THREAD_NUM + 1); i++) {
+        microkit_cothread_semaphore_init(&sem[i]);
+    }
+    
+    // Init file system metadata
+    init_metadata(fs_metadata);
+}
+
+// The notified function requires careful management of the state of the file system
+/*
+  The filesystems should be blockwait for new message if and only if all of working
+  threads are either free(no tasks assigned to them, no pending replies) or blocked in diskio.
+  If the filesystem is blocked here and any working threads are free, then the fatfs_command_queue 
+  must also be empty.
+*/
+void notified(microkit_channel ch) {
+    LOG_FATFS("FS IRQ received::%d\n", ch);
+    if (ch != CLIENT_CH && ch != SERVER_CH) {
+        LOG_FATFS("Unknown channel:%d\n", ch);
+        return;
+    }
+
+    // This variable track if there are new requests being popped from client request queue and pushed into the couroutine pool or not
+    bool new_request_popped = true;
+    // Get the number of elements in the queue is costly so we have a flag defined here to only get the number when needed
+    bool queue_size_init = false;
+
+    uint64_t command_queue_size;
+    uint64_t completion_queue_size;
+
+    uint32_t fs_request_dequeued = 0;
+    uint32_t fs_response_enqueued = 0;
+
+    /**
+      I assume this big while loop is the confusing and critical part for dispatching threads and send back the results.
+    **/
+    while (new_request_popped) {
+        {
+            blk_resp_status_t status;
+            uint16_t success_count;
+            uint32_t id;
+            // Get current element numbers in blk response queue
+            uint32_t len = blk_queue_length_resp(blk_queue_handle);
+
+            while (len > 0) {
+                // This id is the index to the request pool
+                blk_dequeue_resp(blk_queue_handle, &status, &success_count, &id);
```

Comment:
Always check the return value of enqueuing and dequeuing with an assert at least. If it is known to be correct then just an assert is fine.

Having an assert catches a surprising amount of errors for misconfigured setups.

Reply:
There's only one occurrence of `blk_dequeue_resp` that I can find but 8 `blk_enqueue_req` that needs to be checked as well.